File: medium/236. Lowest Common Ancestor of a Binary Tree/236. Lowest Common Ancestor of a Binary Tree.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
    if (!root)
        return null

    if (root.val === p?.val || root.val === q?.val)
        return root

    const leftLCA = lowestCommonAncestor(root.left, p, q)
    const rightLCA = lowestCommonAncestor(root.right, p, q)

    if (leftLCA && rightLCA)
        return root

    return leftLCA || rightLCA
};

File: medium/236. Lowest Common Ancestor of a Binary Tree/236. Lowest Common Ancestor of a Binary Tree.go
package main

/**
 * Definition for a binary tree node.
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	if root.Val == p.Val || root.Val == q.Val {
		return root
	}

	left := lowestCommonAncestor(root.Right, p, q)
	right := lowestCommonAncestor(root.Left, p, q)

	if left != nil && right != nil {
		return root
	} else if left != nil {
        return left
    } else {
        return right
    }
}
